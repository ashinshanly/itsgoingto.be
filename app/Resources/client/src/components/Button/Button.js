import React from 'react'
import PropTypes from 'prop-types'
import EventBus from 'components/EventBus'
import Spinner from 'components/Spinner/Spinner'
import './Button.scss'

class Button extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      disabled : false,
      loading  : false
    }
  }

  componentDidMount = () => {
    this._mounted = true
    this.eventBus = EventBus.getEventBus()
    if (this.props.submitEvent) {
      this.eventListener = this.eventBus.addListener(this.props.submitEvent, () => {
        this.handlePress()
      })
    }
  }

  componentWillUnmount = () => {
    this._mounted = false
    if (this.eventListener) {
      this.eventListener.remove()
    }
  }

  isDisabled = () => this.props.disabled || this.state.disabled

  handlePress = (event) => {
    if (event) {
      event.preventDefault()
    }
    if (!this.isDisabled()) {
      this.setState({
        disabled : true,
        loading  : true
      })
      this.props.callback().then((reset) => {
        if (reset !== false && this._mounted) {
          this.setState({
            disabled : false,
            loading  : false
          })
        }
      })
    }
  }

  render = () => (
    <button
      className={'btn ' + this.props.className + (this.isDisabled() ? ' disabled' : '')}
      disabled={this.props.disabled}
      onClick={this.handlePress}>
      {this.state.loading && <Spinner />}
      {!this.state.loading && <span>{this.props.text}</span>}
    </button>
  )
}

Button.propTypes = {
  text        : PropTypes.string.isRequired,
  className   : PropTypes.string,
  disabled    : PropTypes.bool,
  callback    : PropTypes.func.isRequired,
  submitEvent : PropTypes.string
}

Button.defaultProps = {
  disabled : false,
}

export default Button
