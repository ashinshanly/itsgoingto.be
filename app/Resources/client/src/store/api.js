import { questionSelector, updatePoll } from './poll'
import { answersSelector } from './answers'

// ------------------------------------
// Constants
// ------------------------------------
export const ROUTE_QUESTION = '/api/questions'

// ------------------------------------
// Helpers
// ------------------------------------
export const APIError = (details) => {
  this.name = 'APIError'
  this.details = details
  this.stack = (new Error()).stack
}
APIError.prototype = Object.create(Error.prototype)
APIError.prototype.constructor = APIError

// Extract the json response. Rejects if the response is not 200
export const extractResponse = (response) => {
  if (response.status === 200) {
    return response.json()
  } else {
    return response.json()
    .then((error) => {
      return Promise.reject(
        new APIError({
          status     : response.status,
          statusText : response.statusText,
          error
        })
      )
    })
  }
}

// Parse an error and displays a suitible message to the user.
export const onError = (error) => {
  // TODO : Display an error message to the user
  console.error('There was an error', error.details)
  return false
}

// ------------------------------------
// Actions
// ------------------------------------
export const postPoll = () => (dispatch, getState) =>
  fetch(ROUTE_QUESTION, {
    method: 'POST',
    body: JSON.stringify({
      question: questionSelector(getState()),
      answers: answersSelector(getState()),
    })
  })
  .then(extractResponse)
  .then((response) => dispatch(updatePoll(response)))
  .catch(onError)

export const fetchPoll = (identifier) => (dispatch, getState) =>
  fetch(ROUTE_QUESTION + '/' + identifier)
  .then(extractResponse)
  .then((response) => {
    return dispatch(updatePoll(response))
  })
  .catch(onError)
