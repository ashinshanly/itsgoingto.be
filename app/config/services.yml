# Learn more about services, parameters and containers at
# http://symfony.com/doc/current/book/service_container.html
parameters:
#    parameter_name: value

services:
    session.handler.pdo:
        class:     Symfony\Component\HttpFoundation\Session\Storage\Handler\PdoSessionHandler
        public:    false
        arguments:
            - "mysql:host=%database_host%;dbname=%database_name%"
            - { db_table: sessions, db_username: %database_user%, db_password: %database_password% }

    # authentication failure event listener
    ItsGoingToBe.security.authentication_failure_event_listener:
        class: ItsGoingToBeBundle\Service\AuthenticationListener
        arguments:
            - @doctrine.orm.entity_manager
            - @request_stack
        tags:
            - { name: kernel.event_listener, event: security.authentication.failure, method: onAuthenticationFailure }

    # authentication success event listener
    ItsGoingToBe.security.interactive_login_listener:
        class: ItsGoingToBeBundle\Service\AuthenticationListener
        arguments:
            - @doctrine.orm.entity_manager
            - @request_stack
        tags:
            - { name: kernel.event_listener, event: security.interactive_login, method: onAuthenticationSuccess }