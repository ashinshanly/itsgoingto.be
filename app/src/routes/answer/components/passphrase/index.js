import Passphrase from './passphrase'

export default Passphrase
