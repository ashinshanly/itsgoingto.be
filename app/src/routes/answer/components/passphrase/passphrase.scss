@import "variables";

.passphrase-container {
  z-index: 1050;
  position: fixed;
  width: 90%;
  @media (min-width: $screen-sm-min) {
    width: ($container-sm - ($gutter-width * 8));
  }
  @media (min-width: $screen-md-min) {
    width: ($container-md - ($gutter-width * 10));
  }
  @media (min-width: $screen-lg-min) {
    width: ($container-lg - ($gutter-width * 12));
  }
  transform: translate3d(-50%, -50%, 0px);
  top: 30%;
  left: 50%;
}

.input-passphrase {
  margin-bottom: 1em;

  .input-field {
    font-size: 1.5em;
  }
}
