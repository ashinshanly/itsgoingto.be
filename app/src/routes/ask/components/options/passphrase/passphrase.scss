.input-option-passphrase {
  grid-column: 1 / span 1;

  .input-field.input-field-passphrase {
    font-size: 1.2em;
  }
}
