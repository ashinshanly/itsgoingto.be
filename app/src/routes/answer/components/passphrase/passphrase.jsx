import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { withRouter } from 'react-router'
import { mergeAll } from 'ramda'
import { updatePoll } from 'store/poll'
import { fetchPoll, APIError } from 'store/api'
import { setRequiresPassphrase } from 'store/loader'
import EventBus from 'components/event-bus'
import Button from 'components/button'
import './passphrase.scss'

const KEY_ENTER = 13

class Passphrase extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      error : false
    }
  }

  componentDidMount = () => {
    this.eventBus = EventBus.getEventBus()
  }

  submit = () =>
    this.props.setPassphrase(this.refs.passphrase.value)
      .then(
        () => this.props.fetchPoll(this.props.identifier)
          .then((response) => {
            if (!(response instanceof APIError)) {
              this.props.setRequiresPassphrase(false)
            } else {
              this.setState({
                error : true
              })
            }
          })
      )

  handleKeyPress = (event) => {
    event = event || window.event
    const key = event.keyCode || event.charCode

    if (key === KEY_ENTER) {
      this.eventBus.emit('passphrase-submit')
    }
  }

  render () {
    const { error } = this.state

    return (
      <div className='passphrase-container'>
        <div className={'input-passphrase' + (error ? ' input-error' : '')}>
          <label className='input-label input-label-passphrase' htmlFor='passphrase'>Passphrase</label>
          <input
            className='input-field input-field-passphrase'
            type='text'
            id='passphrase'
            name='passphrase'
            ref='passphrase'
            onKeyDown={this.handleKeyPress} />
          {error &&
            <span className='input-error-label'>
              Passphrase incorrect
            </span>
          }
        </div>
        <Button className='btn pull-right' text='Enter' callback={this.submit} submitEvent='passphrase-submit' />
      </div>
    )
  }
}

Passphrase.propTypes = {
  identifier            : PropTypes.string.isRequired,
  setPassphrase         : PropTypes.func.isRequired,
  fetchPoll             : PropTypes.func.isRequired,
  setRequiresPassphrase : PropTypes.func.isRequired
}

const mapDispatchToProps = (dispatch, props) => ({
  setPassphrase         : (value) => dispatch(updatePoll({ passphrase : value, identifier : props.params.identifier })),
  fetchPoll             : (identifier) => dispatch(fetchPoll(identifier)),
  setRequiresPassphrase : (value) => dispatch(setRequiresPassphrase(value))
})

const mergeProps = (stateProps, dispatchProps, ownProps) => mergeAll([stateProps, dispatchProps, ownProps.params])

export default withRouter(connect(null, mapDispatchToProps, mergeProps)(Passphrase))
