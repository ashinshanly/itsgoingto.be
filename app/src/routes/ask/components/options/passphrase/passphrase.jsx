import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { updatePoll } from 'store/poll/actions'
import './passphrase.scss'

export class Passphrase extends PureComponent {
  handlePassphraseChange = (event) =>
    this.props.updateOptions({
      identifier : '',
      passphrase : event.target.value
    })

  render () {
    const { poll } = this.props

    return (
      <div className='input-option input-option-passphrase'>
        <label
          className='input-label input-label-passphrase'
          htmlFor='passphrase'>
          Set a passphrase (optional)
        </label>
        <input
          className='input-field input-field-passphrase'
          type='text'
          id='passphrase'
          name='passphrase'
          value={poll.passphrase}
          onChange={this.handlePassphraseChange} />
      </div>
    )
  }
}

Passphrase.propTypes = {
  poll          : PropTypes.object.isRequired,
  updateOptions : PropTypes.func.isRequired
}

const mapDispatchToProps = (dispatch) => ({
  updateOptions: (value) => dispatch(updatePoll(value))
})

export default connect(null, mapDispatchToProps)(Passphrase)
