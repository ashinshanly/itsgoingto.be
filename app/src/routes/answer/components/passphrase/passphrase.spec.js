/* global expect */
import React from 'react'
import { shallow } from 'enzyme'
import EventBus from 'components/event-bus'
import Button from 'components/button'
import { APIError } from 'services/api'
import { Passphrase } from './passphrase'

const eventBus = {
  emit: jest.fn(),
  addListener: jest.fn(),
}

EventBus.getEventBus = jest.fn(() => eventBus)

const props = {
  identifier: 'Hd2eJ9Jk',
  setPassphrase: jest.fn(() => Promise.resolve()),
  fetchPoll: jest.fn(() => Promise.resolve()),
  setRequiresPassphrase: jest.fn(() => Promise.resolve()),
}
let wrapper

describe('(Route) answer', () => {
  describe('(Component) passphrase', () => {
    beforeEach(() => {
      wrapper = shallow(<Passphrase {...props} />)
    })

    afterEach(() => {
      jest.clearAllMocks()
    })

    describe('(Action) onKeyDown', () => {
      describe('key is KEY_ENTER', () => {
        it('emits focus request on index + 1', () => {
          wrapper.find('input').simulate('keyDown', { keyCode: 13, preventDefault: () => { } })
          expect(eventBus.emit).toHaveBeenCalledWith('passphrase-submit')
        })
      })
    })

    describe('(Action) onChange', () => {
      it('updates local state', () => {
        wrapper.find('input').simulate('change', {
          target: { value: 'new passphrase' },
        })
        expect(wrapper.state().value).toBe('new passphrase')
      })
    })

    describe('(Action) submit', () => {
      it('should call setPassphrase with the value', () => {
        wrapper.setState({ value: 'passphrase' })
        return wrapper.find(Button).props().callback()
          .then(() => {
            expect(props.setPassphrase).toHaveBeenCalledWith('passphrase')
          })
      })

      it('should call fetchPoll with the identifier', () => {
        return wrapper.find(Button).props().callback()
          .then(() => {
            expect(props.fetchPoll).toHaveBeenCalledWith(props.identifier)
          })
      })

      it('should call setRequiresPassphrase with valid poll response', () => {
        return wrapper.find(Button).props().callback()
          .then(() => {
            expect(props.setRequiresPassphrase).toHaveBeenCalledWith(false)
          })
      })

      it('should set error to true with invalid poll response', () => {
        let fetchPoll = jest.fn(() => Promise.resolve(new APIError('failed')))
        wrapper = shallow(<Passphrase {...props} fetchPoll={fetchPoll} />)
        return wrapper.find(Button).props().callback()
          .then(() => {
            expect(wrapper.state().error).toBe(true)
          })
      })
    })

    describe('(Render)', () => {
      it('matches snapshot', () => {
        expect(wrapper).toMatchSnapshot()
      })

      describe('error is true', () => {
        it('matches snapshot', () => {
          wrapper.setState({ error: true })
          expect(wrapper).toMatchSnapshot()
        })
      })
    })
  })
})