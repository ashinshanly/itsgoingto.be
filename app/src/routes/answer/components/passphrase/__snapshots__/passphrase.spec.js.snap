// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Route) answer (Component) passphrase (Render) error is true matches snapshot 1`] = `
<div
  className="passphrase-container"
>
  <div
    className="input-passphrase input-error"
  >
    <label
      className="input-label input-label-passphrase"
      htmlFor="passphrase"
    >
      Passphrase
    </label>
    <input
      className="input-field input-field-passphrase"
      id="passphrase"
      name="passphrase"
      onChange={[Function]}
      onKeyDown={[Function]}
      type="text"
      value=""
    />
    <span
      className="input-error-label"
    >
      Passphrase incorrect
    </span>
  </div>
  <Button
    callback={[Function]}
    className="btn pull-right"
    disabled={false}
    submitEvent="passphrase-submit"
    text="Enter"
  />
</div>
`;

exports[`(Route) answer (Component) passphrase (Render) matches snapshot 1`] = `
<div
  className="passphrase-container"
>
  <div
    className="input-passphrase"
  >
    <label
      className="input-label input-label-passphrase"
      htmlFor="passphrase"
    >
      Passphrase
    </label>
    <input
      className="input-field input-field-passphrase"
      id="passphrase"
      name="passphrase"
      onChange={[Function]}
      onKeyDown={[Function]}
      type="text"
      value=""
    />
  </div>
  <Button
    callback={[Function]}
    className="btn pull-right"
    disabled={false}
    submitEvent="passphrase-submit"
    text="Enter"
  />
</div>
`;
